libraries/shared-memory-server/src/channel.rs

```diff
@@ -188,11 +189,31 @@ impl ShmemChannel {
     }
 }
 
-fn offsets(server_event_len: usize, client_event_len: usize) -> (usize, usize, usize) {
-    let disconnect_offset = server_event_len + client_event_len;
-    let len_offset = disconnect_offset + mem::size_of::<AtomicBool>();
-    let data_offset = len_offset + mem::size_of::<AtomicU64>();
-    (disconnect_offset, len_offset, data_offset)
+fn offsets(
+    base_ptr: *mut u8,
+    server_event_len: usize,
+    client_event_len: usize,
+) -> (usize, usize, usize) {
+    let (disconnect, len, data) = offset_ptrs(
+        base_ptr
+            .wrapping_add(server_event_len)
+            .wrapping_add(client_event_len),
+    );
+    let base = base_ptr as usize;
+    (
+        disconnect as usize - base,
+        len as usize - base,
+        data as usize - base,
+    )
+}
+
+fn offset_ptrs(next_free: *mut u8) -> (*mut AtomicBool, *mut AtomicU64, *mut u8) {
+    let disconnect_ptr = next_free.wrapping_add(next_free.align_offset(align_of::<AtomicBool>()));
+    let len_ptr = disconnect_ptr
```

Comment:
len_ptr is still misaligned.

The following code works for me:
```rust
    let len_ptr = disconnect_ptr
        .wrapping_add(mem::size_of::<AtomicBool>());
    let len_ptr = len_ptr
        .wrapping_add(len_ptr.align_offset(align_of::<AtomicU64>()));
```

Reply:
Ah yes, of course. Thanks!